src/indexes/text/fuzzy.h

```diff
@@ -1,43 +1,202 @@
-  virtual std::unique_ptr<WordIterator> clone() const override;
-};
+// Fuzzy search using Damerau-Levenshtein distance on RadixTree
+template <typename Target>
+struct FuzzySearch {
+  // Search for words within max_distance edits of pattern
+  // Returns vector of (word, target) pairs
+  static std::vector<std::pair<std::string, Target>> Search(
+      const RadixTree<Target>& tree, 
+      absl::string_view pattern, 
+      size_t max_distance) {
+    
+    std::vector<std::pair<std::string, Target>> results;
+    std::vector<size_t> prev_prev(pattern.length() + 1);  // we need this for transposition
+    std::vector<size_t> prev(pattern.length() + 1);      // we need this for replacemnt and instertion ops
+    std::vector<size_t> curr(pattern.length() + 1);
+    
+    // Initialize first row: cost of inserting pattern characters
+    // For a search string like "race" , the first row would be [0 1 2 3 4]
+    // where each cell represents the cost of charcter insertion to an empty string.
+    for (size_t i = 0; i <= pattern.length(); ++i) {
+      prev[i] = i;
+    }
+    
+    // get path iterator from root
+    auto iter = tree.GetPathIterator("");
+    std::cout << "DEBUG: Starting fuzzy search for pattern='" << pattern << "' max_distance=" << max_distance << std::endl;
+    std::cout << "DEBUG: Root iter.Done()=" << iter.Done() << " iter.CanDescend()=" << iter.CanDescend() << std::endl;
+    if (!iter.Done()) {
+      std::cout << "DEBUG: Root iter.GetPath()='" << iter.GetPath() << "'" << std::endl;
+    }
+    
+    SearchRecursive(iter, pattern, max_distance, "", '\0', 
+                  prev_prev, prev, curr, results);
+    
+    std::cout << "DEBUG: Search complete, found " << results.size() << " results" << std::endl;
+    return results;
+  }
 
-//
-// Takes a root word and iterates over the words that are a fuzzy match
-//
-struct FuzzyWordIterator : public WordIterator {
-  FuzzyWordIterator(indexes::Text& index, absl::string_view word,
-                    size_t distance);
-  virtual bool Done() const override;
-  virtual void Next() override;
-  virtual absl::string_view GetWord() const override;
-  virtual std::unique_ptr<WordIterator> Clone() const override;
-
-  absl::string_view operator*() const { return GetWord(); }
-};
 
+ private:
+static void SearchRecursive(
+    typename RadixTree<Target>::PathIterator iter,
+    absl::string_view pattern, 
+    size_t max_distance,
+    std::string word,  // Current word being built 
+    char prev_char,    // Previous character (for transposition detection)
+    std::vector<size_t>& prev_prev,  // Row i-2 of DP matrix (for transposition)
+    std::vector<size_t>& prev,       // Row i-1 of DP matrix (previous row)
+    std::vector<size_t>& curr,       // Row i of DP matrix (current row being computed)
+    std::vector<std::pair<std::string, Target>>& results) {
+  // ITERATE OVER SIBLINGS at current tree level
+  while (!iter.Done()) {
```

Comment:
We may need hierarchical iterators. I'd guess you want to get a child iterator from from the current node the path iterator is at for this while loop. Otherwise the path iterator would need multiple different "done" states.